docs/guides/spmd_api.md

```diff
@@ -0,0 +1,429 @@
+    return z, None
+
+```
+
+<!-- #region id="1gvf6_a_OeU3" -->
+Now define the `MLP` model as multiple layers of `DotReluDot`.
+
+Note that you can use `nn.scan` and other [lifted transforms](https://flax.readthedocs.io/en/latest/advanced_topics/lift.html) on the module. This will introduce another dimension on the params, and this dimension should be annotated in `metadata_params`. Since the parameters inside `DotReluDot` are already sharded along `model` axis, we don't need to partition multiple layers across `model` dimension here, and therefore denote it as `None`.
+
+To replicate identical layers, you can either use `nn.scan` or a for-loop. `nn.scan` offers faster compilation times, wheras the for-loop is faster on runtime. This guide uses `nn.scan` simply to show that lifted transforms are supported in this API.
+<!-- #endregion -->
+
+```python id="_noFQzrgJ2j0"
+class MLP(nn.Module):
+  num_layers: int
+  depth: int
+  use_scan: bool
+  @nn.compact
+  def __call__(self, x):
+    if self.use_scan:
+      x, _ = nn.scan(DotReluDot, length=self.num_layers, 
+                    variable_axes={"params": 0},
+                    split_rngs={"params": True},
+                    metadata_params={nn.PARTITION_NAME: None}
+                    )(self.depth)(x)
+    else:
+      for i in range(self.num_layers):
+        x, _ = DotReluDot(self.depth)(x)
+    return x
+```
+
+<!-- #region id="pZ6E5btPr5yZ" -->
+# Sharding specification
+
+Now we want to generate the `PartitionSpec` that `pjit` should receive as annotations of input and output data.
+
+The input is relatively simple. For data parallelism, we can shard the batched input `x`s across the `data` axis, by denoting the batch axis as `data`.
+
+<!-- #endregion -->
+
+```python id="63TwLe6vNtID"
+# Input partition rule - data-parallel only
+x_spec = PartitionSpec('data', None)  # dimensions: (batch, length)
+x_spec
+```
+
+<!-- #region id="3d6b15d08980" -->
+However, to generate `PartitionSpec` for the output, we need to use some actual output as reference. One solution is to create a model and evaluate `model.init` abstractly using `jax.eval_shape`, and then use `nn.get_partition_spec` to automatically generate the `PartitionSpec`. 
+
+The code below shows how to get the output spec if you use `flax.training.train_state` to carry out your initialization and training steps, in which case your `pjit`ted function will output a `TrainState`.
+
+> A side note: Here we define our `init_fn` as purely functional and takes `model` and `optimizer` as arguments. This is not necessary - you can simply define with `def init_fn(k, x):` and all will work fine here.
+> 
+> This guide doesn't do it because later we will show you another way to define your model and will run `init_fn` with another model instance. But this does a bit of trouble because `jax.eval_shape` only takes numeric inputs, and we have to create an abstract closure before feeding the function in.
```

Comment:
```suggestion
> This guide doesn't do it because later we will show you another way to define your model and will run `init_fn` with another model instance. However, this is problematic because `jax.eval_shape` only takes numeric inputs, so we have to create an abstract closure before feeding the function in.
```